### PR TITLE
YOLOv5 Export Benchmarks for GPU

### PR DESCRIPTION
**Pull request by glenn-jocher**
Run YOLOv5 benchmarks (speed and accuracy) for all supported export formats. This PR adds GPU benchmarking capability following CPU benchmarking PR https://github.com/ultralytics/yolov5/pull/6613.

Format                      | `export.py --include`         | Model
:---                        | --:                           | :--
[PyTorch](https://pytorch.org/)                     | -                             | `yolov5s.pt`
[TorchScript](https://pytorch.org/docs/stable/jit.html)                 | `torchscript`                 | `yolov5s.torchscript`
[ONNX](https://onnx.ai/)                        | `onnx`                        | `yolov5s.onnx`
[OpenVINO](https://docs.openvino.ai/latest/index.html)                    | `openvino`                    | `yolov5s_openvino_model/`
[TensorRT](https://developer.nvidia.com/tensorrt)                    | `engine`                      | `yolov5s.engine`
[CoreML](https://github.com/apple/coremltools)                      | `coreml`                      | `yolov5s.mlmodel`
[TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model)       | `saved_model`                 | `yolov5s_saved_model/`
[TensorFlow GraphDef](https://www.tensorflow.org/api_docs/python/tf/Graph)         | `pb`                          | `yolov5s.pb`
[TensorFlow Lite](https://www.tensorflow.org/lite)             | `tflite`                      | `yolov5s.tflite`
[TensorFlow Edge TPU](https://coral.ai/docs/edgetpu/models-intro/)         | `edgetpu`                     | `yolov5s_edgetpu.tflite`
[TensorFlow.js](https://www.tensorflow.org/js)               | `tfjs`                        | `yolov5s_web_model/`

Usage: 
```python
git clone https://github.com/ultralytics/yolov5 -b update/bench_gpu  # clone
cd yolov5
pip install -qr requirements.txt coremltools onnx onnxruntime-gpu openvino-dev  # install
pip install -U nvidia-tensorrt --index-url https://pypi.ngc.nvidia.com  # TensorRT

python utils/benchmarks.py --weights yolov5s.pt --img 640 --device 0
```

### Colab Pro+ V100 High-RAM Results

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=0, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CUDA:0 (Tesla V100-SXM2-16GB, 16160MiB)
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 46.7/166.8 GB disk)

Benchmarks complete (458.07s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623                10.19
1             TorchScript        0.4623                 6.85
2                    ONNX        0.4623                14.63
3                OpenVINO           NaN                  NaN
4                TensorRT        0.4617                 1.89
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623                21.28
7     TensorFlow GraphDef        0.4623                21.22
8         TensorFlow Lite           NaN                  NaN
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

### Colab Free T4 Results

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CUDA:0 (Tesla T4, 15110MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 45.2/78.2 GB disk)

Benchmarks complete (501.61s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623                11.51
1             TorchScript        0.4623                 8.70
2                    ONNX        0.4623                17.08
3                OpenVINO           NaN                  NaN
4                TensorRT        0.4628                 3.36
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623                25.13
7     TensorFlow GraphDef        0.4623                25.20
8         TensorFlow Lite           NaN                  NaN
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

### Ultralytics Hyperplane A100

```
benchmarks: weights=yolov5s.pt, imgsz=640, batch_size=1, data=/usr/src/app/data/coco128.yaml, device=, half=False, test=False, pt_only=False
Checking setup...
YOLOv5 🚀 v6.1-174-gc4cb7c6 torch 1.11.0+cu113 CUDA:0 (A100-SXM-80GB, 81251MiB)
Setup complete ✅ (96 CPUs, 1007.7 GB RAM, 1947.5/3519.3 GB disk)

Benchmarks complete (599.41s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4624                 6.83
1             TorchScript        0.4624                 4.63
2                    ONNX        0.4623                43.68
3                OpenVINO           NaN                  NaN
4                TensorRT        0.4629                 1.13
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4624                25.24
7     TensorFlow GraphDef        0.4624                23.77
8         TensorFlow Lite           NaN                  NaN
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

Note: TensorRT exports are fixed at FP16.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Updated model export formats, warmup process tweaks, and benchmarking improvements in the YOLOv5 repository.

### 📊 Key Changes
- Enhanced the export formats DataFrame in `export.py` to include a new column indicating which export formats support GPU acceleration.
- Modified the model warmup method to support a broader range of models, including TensorFlow SavedModel and GraphDef.
- Adjusted the benchmarking script in `utils/benchmarks.py` to utilize device selection and FP16 inference if available and supported.

### 🎯 Purpose & Impact
- **GPU Acceleration Visibility**: Users can now easily identify which export formats can take advantage of GPU acceleration, leading to more informed decisions on export formats based on their deployment needs.
- **Enhanced Warmup**: Updated model warmup process that now accounts for more model types, potentially leading to more efficient and reliable model performance post-export.
- **Benchmarking Customization and Accuracy**: With added device and precision flags, users can run benchmarks that more closely represent their production environments, yielding more accurate performance measurements. 

These changes aim to improve the user experience in model exporting and benchmarking, ensuring users can optimize their models based on the specific capabilities of their hardware.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@zldrobit I'm adding GPU support to utils/benchmarks.py. 

I haven't tried to update our TF models for GPU inference support yet though. Which models should be able to support TF GPU inference? SavedModel, GraphDef and TFLite?

**Comment by zldrobit** [comment]
@glenn-jocher SavedModel and GraphDef both support GPU inference, while TFLite GPU inference is not supported by PyPI package.  TFLite GPU inference on X86 requires building TensorFlow from source.

**Comment by glenn-jocher** [comment]
@zldrobit got it! Yes now I see that SavedModel and GraphDef are both running correctly on GPU. I've added updates and now they are appearing correctly when benchmarking.

**Comment by glenn-jocher** [comment]
@zldrobit updated results including TF models:

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=0, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CUDA:0 (Tesla V100-SXM2-16GB, 16160MiB)
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 46.7/166.8 GB disk)

Benchmarks complete (458.07s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623                10.19
1             TorchScript        0.4623                 6.85
2                    ONNX        0.4623                14.63
3                OpenVINO           NaN                  NaN
4                TensorRT        0.4617                 1.89
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623                21.28
7     TensorFlow GraphDef        0.4623                21.22
8         TensorFlow Lite           NaN                  NaN
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

**Comment by throughheavenandearthimtheonlyone** [comment]
Hello,
Goodday

I have a temporal detection model using YOLOv5. Can you help me convert it to ONNX or TensorRT?

**Comment by pderrenger** [comment]
Hello,  

To convert your YOLOv5 model to ONNX or TensorRT, you can use the `export.py` script provided in the repository. For example:  

```bash  
python export.py --weights your_model.pt --include onnx engine  
```  

Make sure to install the required dependencies from the [Export Guide](https://docs.ultralytics.com/yolov5/tutorials/model_export/). Let us know if you encounter any issues!
